src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -844,16 +844,95 @@ public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node
             Debug.Assert(!IsConditionalState);
             LearnFromAnyNullPatterns(node.Expression, node.Pattern);
             VisitPatternForRewriting(node.Pattern);
-            Visit(node.Expression);
-            var expressionState = ResultType;
-            var state = PossiblyConditionalState.Create(this);
-            var labelStateMap = LearnFromDecisionDag(node.Syntax, node.DecisionDag, node.Expression, expressionState, ref state);
-            var trueState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenFalseLabel : node.WhenTrueLabel, out var s1) ? s1.state : UnreachableState();
-            var falseState = labelStateMap.TryGetValue(node.IsNegated ? node.WhenTrueLabel : node.WhenFalseLabel, out var s2) ? s2.state : UnreachableState();
-            labelStateMap.Free();
-            SetConditionalState(trueState, falseState);
+            if (!VisitPossibleConditionalAccess(node.Expression, out var conditionalStateWhenNotNull)
+                || !trySetStateWhenNotNull(node.Pattern, conditionalStateWhenNotNull))
```

Comment:
Don't we need to visit the pattern (`LearnFromDecisionDag` below) in all cases, to set the types of variables (recorded in `_variables`)?

Reply:
FYI I'm investigating the issue with top-level `var`.

Reply:
I've changed the implementation to pass 'stateWhenNotNull' into 'LearnFromDecisionDag' and use it where appropriate instead.